ingest/expression_files/mtx.py

```diff
@@ -183,64 +194,96 @@ def transform(self):
                 if not is_sorted:
                     raise ValueError("MTX file must be sorted")
                 else:
+
                     visited_expression_idx.append(current_idx)
-                    # Create data arrays from prior gene
+                    # Add current gene's gene model
+                    # Data array for cell names
                     if last_idx != 0:
-                        # Data array for cell names
-                        da_cells = GeneExpression.create_data_arrays(
-                            name=gene,
-                            array_type=f"{gene} Cells",
+                        last_gene_id, last_gene = self.genes[current_idx - 2].split(
+                            "\t"
+                        )
+                        current_data_arrays = []
+                        # Data arrays for cells
+                        for data_array in GeneExpression.create_data_arrays(
+                            name=f"{last_gene} Cells",
+                            array_type="cells",
                             values=exp_cells,
-                            linear_data_type="Study",
+                            linear_data_type="Gene",
                             linear_data_id=model_id,
                             **self.data_array_kwargs,
-                        )
-                        data_arrays.extend(da_cells)
-                        # Data array for expression values
-                        da_exp = GeneExpression.create_data_arrays(
-                            name=gene,
-                            array_type=f"{gene} Expression",
+                        ):
+                            current_data_arrays.append(data_array)
+                        # Data arrays for expression values
+                        for data_array in GeneExpression.create_data_arrays(
+                            name=f"{last_gene} Expression",
+                            array_type=f"expression",
                             values=exp_scores,
                             linear_data_type="Gene",
                             linear_data_id=model_id,
                             **self.data_array_kwargs,
-                        )
-                        data_arrays.extend(da_exp)
+                        ):
+                            current_data_arrays.append(data_array)
+                        # Determine if models should be batched
+                        if (
+                            len(data_arrays) + len(current_data_arrays)
+                            >= GeneExpression.DATA_ARRAY_BATCH_SIZE
+                        ):
+                            yield gene_models, GeneExpression.COLLECTION_NAME
+                            yield data_arrays, DataArray.COLLECTION_NAME
+                            num_processed += len(gene_models)
```

Comment:
This code seems to duplicate that [around line 282 in `dense_ingestor.py`](https://github.com/broadinstitute/scp-ingest-pipeline/pull/132/files#diff-648fded1b7ef87ee4e8498e7d0346c28R286).

If the duplication is feasible to abstract, could you do so?

If not, could you apply my refinement for "Easier reading" from other there to here?

Reply:
This piece of code has been refactored